pythia/models/butd.py

```diff
@@ -119,41 +119,36 @@ def forward(self, sample_list):
             dtype=torch.float,
         )
 
-        # For beam search inference. Currently beam seach for BUTD works only 
-        # with batch_size = 1 and should be used with run_type inference only.
-        # TODO : Implement batch beam search
-        if self.config["inference"]["type"] == "beam_search":
-            beam_search = BeamSearch(
+        if self.config["inference"]["type"] == "greedy":
+            decoder = BeamSearch(self.vocab, 1)
+        elif self.config["inference"]["type"] == "beam_search":
+            decoder = BeamSearch(
                 self.vocab, self.config["inference"]["params"]["beam_length"]
             )
-            sample_list = beam_search.init_batch(sample_list)
+        elif self.config["inference"]["type"] == "nucleus_sampling":
+            decoder = NucleusSampling(
+                self.vocab, self.config["inference"]["params"]["sum_threshold"]
+            )
 
+        sample_list = decoder.init_batch(sample_list)
         batch_size = sample_list.image_feature_0.size(0)
         data, sample_list, timesteps = self.prepare_data(sample_list, batch_size)
         output = None
         batch_size_t = batch_size
         for t in range(timesteps):
             data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
-            if self.config["inference"]["type"] == "beam_search":
-                pi_t = data["texts"]
-            else:
-                pi_t = data["texts"][:, t].unsqueeze(-1)
+            pi_t = data["texts"]
```

Comment:
Amm, the else condition is completely gone here. We don't require it now?

Reply:
Nope. The else was required for greedy when it had its own code. But now since we're using Beam Search with beam_size = 1 for Greedy this else was no longer needed.